src/Storages/MergeTree/ReplicatedMergeTreeMergeStrategyPicker.cpp

```diff
@@ -0,0 +1,152 @@
+// but that doesn't have a significant impact (in one case it will wait for the execute_merges_on_single_replica_time_threshold,
+// in another just 2 replicas will do the merge)
+std::optional<std::string> ReplicatedMergeTreeMergeStrategyPicker::pickReplicaToExecuteMerge(const ReplicatedMergeTreeLogEntryData & entry)
+{
+    time_t threshold = execute_merges_on_single_replica_time_threshold;
+    if (
+        threshold == 0 // feature turned off
+        || entry.type != ReplicatedMergeTreeLogEntry::MERGE_PARTS // it is not a merge log entry
+       )
+        return std::nullopt;
+
+    auto now = time(nullptr);
+
+    if (entry.create_time + threshold <= now) // too much time waited
+        return std::nullopt;
+
+    if (now - last_refresh_time > REFRESH_STATE_MAXIMUM_INTERVAL_SECONDS) // last state refresh was too long ago, need to sync up the replicas list
+        refreshState();
+
+    auto hash = getEntryHash(entry);
+
+    std::lock_guard lock(mutex);
+
+    auto num_replicas = active_replicas.size();
+
+    if (num_replicas==0)
+        return std::nullopt;
+
+    auto replica_index = static_cast<int>(hash % num_replicas);
+
+    if (replica_index == current_replica_index)
+        return std::nullopt;
+
+    return active_replicas.at(replica_index);
+}
+
+
+void ReplicatedMergeTreeMergeStrategyPicker::refreshState()
+{
+    auto threshold = storage.getSettings()->execute_merges_on_single_replica_time_threshold.totalSeconds();
+
+    auto now = time(nullptr);
+
+    // the setting is not changed, and last state refresh was done recently
+    if (threshold == execute_merges_on_single_replica_time_threshold
+        && now - last_refresh_time < REFRESH_STATE_MINIMUM_INTERVAL_SECONDS)
+        return;
+
+    if (threshold == 0)
+    {
+        // we can reset the settings w/o lock (it's atomic)
+        execute_merges_on_single_replica_time_threshold = threshold;
+        return;
+    }
+
+    auto zookeeper = storage.getZooKeeper();
+    auto all_replicas = zookeeper->getChildren(storage.zookeeper_path + "/replicas");
+
+    // TODO: do we need that sort or we can rely that zookeeper will return them in deterministic order?
+    std::sort(all_replicas.begin(), all_replicas.end());
+
+    std::vector<String> active_replicas_tmp;
+    int current_replica_index_tmp = -1;
+
+    for (const String & replica : all_replicas)
+    {
+        if (zookeeper->exists(storage.zookeeper_path + "/replicas/" + replica + "/is_active"))
+        {
+            active_replicas_tmp.push_back(replica);
+            if (replica == storage.replica_name)
+            {
+                current_replica_index_tmp = active_replicas_tmp.size() - 1;
+            }
+        }
+    }
+
+    if (current_replica_index_tmp < 0 || active_replicas_tmp.size() < 2)
+    {
+        LOG_ERROR(storage.log, "Can't find current replica in the active replicas list, or too few active replicas to use execute_merges_on_single_replica_time_threshold!");
```

Comment:
Mmm, but why it should be logged as `LOG_ERROR`? Doesn't seem like it's extraordinary situation.

Reply:
changing to WARNING.